modules/axtask/src/run_queue.rs

```diff
@@ -75,54 +234,111 @@ impl AxRunQueue {
             can_preempt
         );
         if can_preempt {
-            self.resched(true);
+            self.inner.resched(true);
         } else {
             curr.set_preempt_pending(true);
         }
     }
 
+    /// Exit the current task with the specified exit code.
+    /// This function will never return.
     pub fn exit_current(&mut self, exit_code: i32) -> ! {
         let curr = crate::current();
         debug!("task exit: {}, exit_code={}", curr.id_name(), exit_code);
-        assert!(curr.is_running());
+        assert!(curr.is_running(), "task is not running: {:?}", curr.state());
         assert!(!curr.is_idle());
         if curr.is_init() {
-            EXITED_TASKS.lock().clear();
+            EXITED_TASKS.with_current(|exited_tasks| exited_tasks.clear());
             axhal::misc::terminate();
         } else {
             curr.set_state(TaskState::Exited);
-            curr.notify_exit(exit_code, self);
-            EXITED_TASKS.lock().push_back(curr.clone());
-            WAIT_FOR_EXIT.notify_one_locked(false, self);
-            self.resched(false);
+
+            // Notify the joiner task.
+            curr.notify_exit(exit_code);
+
+            // Push current task to the `EXITED_TASKS` list, which will be consumed by the GC task.
+            EXITED_TASKS.with_current(|exited_tasks| exited_tasks.push_back(curr.clone()));
+            // Wake up the GC task to drop the exited tasks.
+            WAIT_FOR_EXIT.with_current(|wq| wq.notify_one(false));
+            // Schedule to next task.
+            self.inner.resched(false);
         }
         unreachable!("task exited!");
     }
 
-    pub fn block_current<F>(&mut self, wait_queue_push: F)
-    where
-        F: FnOnce(AxTaskRef),
-    {
+    /// Block the current task, put current task into the wait queue and reschedule.
+    /// Mark the state of current task as `Blocked`, set the `in_wait_queue` flag as true.
+    /// Note:
+    ///     1. The caller must hold the lock of the wait queue.
+    ///     2. The caller must ensure that the current task is in the running state.
+    ///     3. The caller must ensure that the current task is not the idle task.
+    ///     4. The lock of the wait queue will be released explicitly after current task is pushed into it.
+    pub fn blocked_resched(&mut self, mut wq_guard: WaitQueueGuard) {
```

Comment:
这个接口我认为不应该有 `wq_guard: WaitQueueGuard ` 这个入参，原因有几个
1、  `block` 应该是更宽泛的调度接口，而不仅仅是 `block on wait queue`
2、 `wait queue` 的相关动作(入队) 为什么不放在`waitqueue.rs` 里面实现呢？`lock guard` 跨文件和模块不是很好，而且这里确实可以放在它自己模块处理

Reply:
See this comment https://github.com/arceos-org/arceos/pull/176#discussion_r1788855228

 we have to make sure task's state is set as Blocked with holding the lock of wait queue.
 
 While @equation314 suggested that we should put all task state manipulation to `run_queue.rs`

Besides, instead of "block on wait queue`, where should a task be if it is not running or ready

Reply:
**Regarding whether there will be other blocking situations**
I mean, are there other synchronization mechanisms (futex wait, completion wait,...) that are also blocking, not blocking on the waitqueue data structure, but blocking on some other queues
**Regarding the issue of whether the state needs to be locked**
I temporarily agree that the state machine settings need to be unified somewhere (whether it is really necessary, Linux seems to be more casual). The key purpose of doing this here is that the change and judgment of the state need to be protected by a lock at the same time. The logic of the wait queue should be reused here, so if the state change has its own lock, isn't it better?

Reply:
1. For arceos, futex wait, completion wait can all implemented based on wait queue (like Mutex in ArceOS's api).

2. modifications to the state may require locking at different granularities, such as from the scheduler or wait queue. Some state modifications can be lock-free, such as changes to the current task's state in `yield_current` or `sleep_until`.

Reply:
Here is an implementation of a synchronization mechanism. I don’t quite remember why it was implemented this way, but what is certain is that waitque is definitely not a simple data structure with the smallest granularity.
https://github.com/Starry-OS/axsync.git ；
Of course, I won’t dwell on this too much, because it may not be wise to think too much about extensibility for features that don’t exist or haven’t been implemented yet; but I want everyone to know that this is possible.

Reply:
> but what is certain is that waitque is definitely not a simple data structure with the smallest granularity.
https://github.com/Starry-OS/axsync.git ；

So what is the difference between `WaitTaskList` and `WaitQueue` besides their data structure.